Fix #2532: matrix index symbol `end` not working when used inside a sub-expression

josdejong opened the pull request:
Fix #2532. Instead of only supporting the case where the index is `end`, it now handles any sub-expression inside with `end` is used.

@dvd101x or @gwhitney can you have a look at this PR?

dvd101x reviewed (approved):
For me it looks all ok.

gwhitney commented:
Looks fine to me. I frankly did not know about `end` until seeing the associated issue. But this seems to work well. My only comments are not directly related to this change: 

1) When 'end' occurs inside an accessor to an object, a somewhat strange error is thrown. I am not surprised it's a runtime exception, as I am not sure what `end` would mean (although I suppose it could mean the last key in the order JavaScript presents them in `Object.keys()`) but it might be clearer if the message were something like "`end` only valid in array/matrix indices." 

(2) In nested indexing `a[b[end]]` the `end` of course means the last index of `b`, as it should, but there is no way of getting the last index of `a` in the innermost context. Not sure that you would want/need a way (I mean, you can already write `a[b[size(a)[1]]]`) but perhaps this should be pointed out in the documentation somewhere and/or be tested in a unit test. 

The two points interact slightly in that if I set `a`  to `[-1,-2,-3]` and `b` to `{3:2}` I half expected `a[b[end]]` to produce -2, in that I thought object indexing would not set "end" so the "end" value from array-indexing `a` would still be valid in the inner context and be used as the key for `b` to produce 2, which would then be used to index `a`.

Not sure it would make any sense to address these points here; but they also seemed so minor that I wasn't sure they were even worth opening an issue for; so I thought I would just compromise and mention them here to see if you thought they were worth doing anything about and if so in what context (as additions to this PR or somewhere separate).

dvd101x commented:
I don't recall seeing [end] in the documentation, I just tested something similar to what [Matlab/Octave does](https://www.mathworks.com/help/matlab/matlab_oop/object-end-indexing.html) and also found that it was being used on [Rocket trajectory optimization](https://mathjs.org/examples/browser/rocket_trajectory_optimization.html.html)

Maybe another issue would be to add the [end] indexing into the documentation for [matrices](https://mathjs.org/docs/datatypes/matrices.html). If it's not documented yet I can work on it.

gwhitney commented:
It is mentioned in `docs/expressions/syntax.md`.

dvd101x commented:
Thanks Glen!

I’m sorry about missing that.

josdejong commented:
Thanks for the feedbacks, I've addressed them. Can you have another brief look @gwhitney ?

1. I noticed that half of the logic in `IndexNode.prototype._compile` is redundant, so I removed that 😁 
2. I've described the behavior of `end` in case of nested documentation
3. A proper error is thrown now when trying to use `end` on an object instead of matrix, like:
    ```
    TypeError: Cannot resolve "end": context must be a Matrix, Array, or string but is Object
    ```

gwhitney commented:
Other than possibly adding a unit test that in a nested occurrence `end` returns the last index of the innermost array (up to you if you think such a test is worth it) all looks good to me. Let me know if you want me to merge this or if you will.
